tests/xharness/Execution/MLaunchArguments.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Xharness.Execution {
+	// mlaunch is really important and has a lot of arguments that are known but
+	// used to be passed as strings. This class allows to add arguments without
+	// knowing the exact string and will also validate that an argument that
+	// needs a value does contain the valuw
+	public enum MlaunchArgumentType {
+		SdkRoot,
+		ListDev,
+		ListSim,
+		OutputFormat,
+		ListExtraData,
+	}
+
+	struct MlaunchArgument {
+		public MlaunchArgumentType Type { get; set; }
+		public string Value { get; set; }
+
+		public MlaunchArgument ((MlaunchArgumentType type, string value) data)
+		{
+			Type = data.type;
+			Value = data.value;
+			if (!IsValid (out var reason)) 
+				throw new ArgumentException (reason);
+		}
+
+		public MlaunchArgument (MlaunchArgumentType type)
+		{
```

Comment:
To be honest, it would be better to have a class for each argument class accepting exactly what it needs as value (maybe there's none, maybe it needs an array, maybe just a string..)

There would be `abstract MlaunchArgument` class and the argument classes would implement some `ToCommandLine` method. Validation would happen in ctor of each of those.

It might also help in future because instead of providing the tuple (Type, Value) you would have the value named so you know what it is, something like:
`new ListSimArgument(deviceId: value)` whereas now you actually cannot tell what the argument is (is it device ID, is it iOS version...)

But again, no hard feelings around this, just the only thing I mind is that now the code lets you create invalid argument on compile time by letting you (not) supply value.

Reply:
I can see now in other usages that we might benefit from this more by having specialized constructors such as the `-argument=-app-arg:-something` one

Reply:
yes. this was a first implementation. I think it will evolve until we find an API that will cover all the cases. At least is much more secure than using strings.